File: CompetitiveProgramming/DMOJ/Uncategorized/Postfix Notation/2021-08-11/PostfixNotation.py
"""
NOTES

5 5 + 6 * 8 -
52.0

((5 + 5) * 6) - 8

if operator
a = pop()
b = pop()
b operator a


* / + - % ^

"""

import sys

def main():
    input = sys.stdin.readline

    data = input().split()
    stack = []
    for d in data:
        if d.isnumeric():
            stack.append(int(d))
        else:
            b = stack.pop()
            a = stack.pop()
            if d == "*":
                stack.append(a*b)
            elif d == '/':
                stack.append(a/b)
            elif d == '+':
                stack.append(a+b)
            elif d == '-':
                stack.append(a-b)
            elif d == '%':
                stack.append(a%b)
            else:
                stack.append(a**b)
    print(stack[0])
main()

File: CompetitiveProgramming/DMOJ/Uncategorized/Postfix Notation/2021-08-18/PostfixNotation.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Stack;

public class PostfixNotation
{
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    String[] input()
    {
        try
        {
            return br.readLine().split(" ");
        }
        catch(IOException ioe)
        {
            ioe.printStackTrace();
        }
        return null;
    }

    boolean isNum(String s)
    {
        try
        {
            Double.parseDouble(s);
        }
        catch(NumberFormatException nfe)
        {
            return false;
        }
        return true;
    }

    void run()
    {
        String[] data = input();
        Stack<Double> s = new Stack<>();
        for(String d : data)
        {
            if(isNum(d))
            {
                s.push(Double.parseDouble(d));
            }
            else
            {
                double b = s.pop();
                double a = s.pop();
                switch(d.charAt(0))
                {
                    case '+': s.push(a+b); break;
                    case '-': s.push(a-b); break;
                    case '*': s.push(a*b); break;
                    case '/': s.push(a/b); break;
                    case '%': s.push(a%b); break;
                    case '^': s.push(Math.pow(a, b)); break;
                }
            }
        }
        System.out.println(s.pop());
    }

    public static void main(String[] args)
    {
        PostfixNotation pn = new PostfixNotation();
        pn.run();
    }
}